### PR TITLE
JDK-8293701: jdeps InverseDepsAnalyzer runs into NoSuchElementException: No value present

### PR DESCRIPTION
**Pull request by MBaesken**
We noticed that with certain jar file input, jdeps runs into the following exception, this happens with jdk11, 17 and 20.

jdeps.exe --multi-release 11 --module-path . --inverse --package com.sap.nw.performance.supa.client test.jar

Inverse transitive dependences matching packages [com.sap.nw.performance.supa.client]
Exception in thread "main" java.util.NoSuchElementException: No value present
        at java.base/java.util.Optional.get(Optional.java:148)
        at jdk.jdeps/com.sun.tools.jdeps.InverseDepsAnalyzer.lambda$inverseDependences$2(InverseDepsAnalyzer.java:150)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)

So an additional  check might be a good idea.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293701](https://bugs.openjdk.org/browse/JDK-8293701): jdeps InverseDepsAnalyzer runs into NoSuchElementException: No value present

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10300/head:pull/10300` \
`$ git checkout pull/10300`

Update a local copy of the PR: \
`$ git checkout pull/10300` \
`$ git pull https://git.openjdk.org/jdk pull/10300/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10300`

View PR using the GUI difftool: \
`$ git pr show -t 10300`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10300.diff">https://git.openjdk.org/jdk/pull/10300.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
Do you think it would be possible to create a small test case that demonstrates the issue? A number of the fixes to jdeps in recent releases have be due to some unusual scenarios and I think it would be useful to build up more test cases for this scenarios, otherwise they could easily break again if there are no tests.

**Comment by jaikiran** [comment]
In addition to what Alan stated, it seems odd that a `isPresent` check is (now) necessary when the previously line states that the module "must be present". So I think having a reproducer, like Alan noted, would be important to decide what the fix should be.

**Comment by MBaesken** [comment]
> Do you think it would be possible to create a small test case that demonstrates the issue? A number of the fixes to jdeps in recent releases have be due to some unusual scenarios and I think it would be useful to build up more test cases for these scenarios, otherwise they could easily break again if there are no tests.

Hi Alan and Jaikiran , I agree  a small test case would be beneficial .  So far I unfortunately have only a large reproducer that cannot be shared and comes in the format of a couple of jars - far from ideal .

**Comment by mlchung** [comment]
What is the cause of this issue?   It expects the dependency can be found but missing in this case.   I wonder if it's due to a `requires static` dependency?

**Comment by MBaesken** [comment]
> What is the cause of this issue? It expects the dependency can be found but missing in this case. I wonder if it's due to a `requires static` dependency?

Hi Mandy, good question, so far I have just a zip with a coupple of jars inside , still try to find someone who can point me to the sources .

**Comment by AlanBateman** [comment]
> > What is the cause of this issue? It expects the dependency can be found but missing in this case. I wonder if it's due to a `requires static` dependency?
> 
> Hi Mandy, good question, so far I have just a zip with a coupple of jars inside , still try to find someone who can point me to the sources .

Could you use `jar --file=<file> --describe-module` to get a describe of each module? The bug report doesn't need the names of the SAP modules but getting a picture of what the input to jdeps is would help figure out what this is about.

**Comment by MBaesken** [comment]
Hi Alan, Mandy so far I unfortunately do not have the sources.

Here are as far as I can see the most important jars, with module descriptions.
I wondered about the com.sap.nw.performance.commons  requires - I did not notice and export for this one, looked a bit strange to me (this should have been in performance-commons.jar at least the authors told me so ...

supa_jdeps/SUPA/lib> jdk/images/jdk/bin/jar --file=supa-client.jar  --describe-module
com.sap.nw.performance.supa.client jar:file:///supa_jdeps/SUPA/lib/supa-client.jar!/module-info.class
exports com.sap.nw.performance.supa.automation
requires com.sap.nw.performance.commons transitive
requires com.sap.nw.performance.supa.api transitive
requires com.sap.nw.performance.supa.base transitive
requires java.base mandated
requires java.logging transitive
qualified exports com.sap.nw.performance.supa.automation.implementation to supa.tests
opens com.sap.nw.performance.supa.automation


supa_jdeps/SUPA/lib> jdk/images/jdk/bin/jar --file=supa-base.jar  --describe-module
com.sap.nw.performance.supa.base jar:file:///test/SUPA/lib/supa-base.jar!/module-info.class
exports com.sap.nw.performance.supa.base
exports com.sap.nw.performance.supa.base.command
exports com.sap.nw.performance.supa.base.config
exports com.sap.nw.performance.supa.base.cpuinfo
exports com.sap.nw.performance.supa.base.http
exports com.sap.nw.performance.supa.base.io
exports com.sap.nw.performance.supa.base.processinfo
exports com.sap.nw.performance.supa.base.processinfo.os
exports com.sap.nw.performance.supa.base.system
exports com.sap.nw.performance.supa.base.time
exports com.sap.nw.performance.supa.base.utils
exports com.sap.nw.performance.supa.base.utils.file
requires JNativeHook transitive
requires com.github.spotbugs.annotations static
requires com.google.common transitive
requires com.google.gson transitive
requires com.sap.nw.performance.commons transitive
requires com.sap.nw.performance.supa.api transitive
requires ecs transitive
requires java.base mandated
requires java.desktop transitive
requires java.net.http transitive
requires org.apache.commons.collections4 transitive
requires org.apache.commons.text transitive
requires org.apache.httpcomponents.httpclient transitive
requires org.apache.httpcomponents.httpcore transitive
requires org.apache.httpcomponents.httpmime transitive
qualified exports com.sap.nw.performance.supa.base.cpuinfo.os to supa.tests


supa_jdeps/SUPA/lib> jdk/images/jdk/bin/jar --file=supa-api.jar  --describe-module
com.sap.nw.performance.supa.api jar:file:///supa_jdeps/SUPA/lib/supa-api.jar!/module-info.class
exports com.sap.nw.performance.supa.api
exports com.sap.nw.performance.supa.api.config
exports com.sap.nw.performance.supa.api.manager.state
exports com.sap.nw.performance.supa.api.processinfo
exports com.sap.nw.performance.supa.api.processprofiler
requires com.sap.nw.performance.commons transitive
requires java.base mandated

supa_jdeps/SUPA/lib> jdk/images/jdk/bin/jar --file=performance-commons.jar  --describe-module
com.sap.nw.performance.commons jar:file:///supa_jdeps/SUPA/lib/performance-commons.jar!/module-info.class
exports com.sap.nw.performance.commons.command
exports com.sap.nw.performance.commons.config
exports com.sap.nw.performance.commons.exceptions
exports com.sap.nw.performance.commons.outlier
exports com.sap.nw.performance.commons.remote
exports com.sap.nw.performance.commons.remote.command
exports com.sap.nw.performance.commons.results
exports com.sap.nw.performance.commons.utils
exports com.sap.nw.performance.commons.utils.html
exports com.sap.nw.performance.commons.utils.http
exports com.sap.nw.performance.commons.utils.jdsr
exports com.sap.nw.performance.commons.utils.net
exports com.sap.nw.performance.commons.utils.path
exports com.sap.nw.performance.commons.utils.zip
exports com.sap.nw.performance.limits
exports com.sap.nw.performance.measrules
exports com.sap.nw.performance.measrules.rule
requires client.combined
requires com.github.spotbugs.annotations static
requires com.google.gson
requires commons.cli transitive
requires commons.configuration transitive
requires commons.lang transitive
requires java.base mandated
requires java.desktop
requires java.management
requires java.scripting
requires java.sql transitive
requires junit static
requires mustache.compiler
requires opencsv
requires org.apache.commons.io transitive
requires org.apache.commons.lang3
requires org.apache.commons.text
requires org.apache.derby.client static
requires org.apache.derby.engine static
requires org.apache.derby.tools static
requires org.objectweb.asm
requires org.objectweb.asm.commons
requires org.objectweb.asm.tree
requires org.objectweb.asm.tree.analysis
requires org.objectweb.asm.util
requires org.openjdk.nashorn
opens com.sap.nw.performance.measrules

**Comment by MBaesken** [comment]
One additional comment about the jdeps --help output 
https://github.com/openjdk/jdk/blob/master/src/jdk.jdeps/share/classes/com/sun/tools/jdeps/resources/jdeps.properties#L139


main.opt.I=\
\  -I\n\
\  --inverse                     Analyzes the dependences per other given options\n\
\                                and then find all artifacts that directly\n\
\                                and indirectly depend on the matching nodes.\n\
\                                This is equivalent to the inverse of\n\
\                                compile-time view analysis and print\n\
\                                dependency summary.  This option must use\n\
\                                with --require, --package or --regex option.

Should this be better  'This option must be used . . .' ?  And what about a sentence explaining what 'nodes' you are talking about ?

Btw is there some trace mode telling me more where exactly the jdeps tool fails in this particular case ?

**Comment by mlchung** [comment]
@MBaesken I move the debugging discussion to the JBS issue (https://bugs.openjdk.org/browse/JDK-8293701).   We should resume this PR once the cause of the problem is understood.

**Comment by mlchung** [comment]
I created a reproducible test case attached in the JBS issue.   This is caused by the optional dependency `requires static M;`  which is not resolved in the runtime configuration even if it's observable.   

```
diff --git a/src/jdk.jdeps/share/classes/com/sun/tools/jdeps/InverseDepsAnalyzer.java b/src/jdk.jdeps/share/classes/com/sun/tools/jdeps/InverseDepsAnalyzer.java
index 5720fe7f60a..18ef9af3422 100644
--- a/src/jdk.jdeps/share/classes/com/sun/tools/jdeps/InverseDepsAnalyzer.java
+++ b/src/jdk.jdeps/share/classes/com/sun/tools/jdeps/InverseDepsAnalyzer.java
@@ -145,6 +145,9 @@ public class InverseDepsAnalyzer extends DepsAnalyzer {
                 .forEach(m -> {
                     builder.addNode(m);
                     m.descriptor().requires().stream()
+                        // filter "requires static" if the module is not resolved in the configuration
+                        .filter(req -> !req.modifiers().contains(Requires.Modifier.STATIC)
+                                           || configuration.findModule(req.name()).isPresent())
                         .map(Requires::name)
                         .map(configuration::findModule)  // must be present
                         .forEach(v -> builder.addEdge(v.get(), m));
```

**Comment by MBaesken** [comment]
Hi Mandy,

> This is caused by the optional dependency requires static M; which is not resolved in the runtime configuration even if it's observable.

looks like this is indeed the issue. With your patch applied  the exception  "Exception in thread "main" java.util.NoSuchElementException: No value present"  is gone .

I reproduced it on my large example as well.
Reason was that some build dependency jar(s)  were missing at runtime,  and jdeps did not accept this.

File: src/jdk.jdeps/share/classes/com/sun/tools/jdeps/InverseDepsAnalyzer.java
Comment (mlchung) [inline src/jdk.jdeps/share/classes/com/sun/tools/jdeps/InverseDepsAnalyzer.java:150]: Nit: indent further to the right for readability (maybe 16 spaces)

File: test/langtools/tools/jdeps/optionalDependency/OptionalDependencyError.java
Comment (mlchung) [inline test/langtools/tools/jdeps/optionalDependency/OptionalDependencyError.java:43]: I'd rename the test to `OptionalDependencyTest` since it's not expecting an error.

File: test/langtools/tools/jdeps/optionalDependency/OptionalDependencyError.java
Comment (mlchung) [inline test/langtools/tools/jdeps/optionalDependency/OptionalDependencyError.java:73]: This test case verifies when m3 is resolved in the configuration.   I suggest to rename this to `optionalDependenceResolved`.  Please also add the comment something like

```
/*
 * Test if a requires static dependence is resolved in the configuration.
 */
```

can you add one more test case that m3 is not resolved `optinalDependenceNotResolved`?

**Comment by MBaesken** [comment]
Hi Mandy, I adjusted the indentation and renamed and adjusted the test.
Especially I removed the  --add-modules m3  from the test because it is not necessary any more.

File: test/langtools/tools/jdeps/optionalDependency/OptionalDependencyTest.java
Comment (mlchung) [inline test/langtools/tools/jdeps/optionalDependency/OptionalDependencyTest.java:73]: Since you removed `--add-modules m3` is removed, this test case is `OptionalDependenceNotResolved`.    Although the bug only happens if m3 is not resolved in the configuration,  a test case to resolve m3 (with `--add-modules m3`) is for completeness and also helps future debugging in case any of these cases was broken.

**Comment by MBaesken** [comment]
Hi Mandy, thanks for the advice , I enhanced the test.

**Review by mlchung — APPROVED** [review]
Looks good.  Thanks for fixing this.

**Comment by MBaesken** [comment]
> Looks good. Thanks for fixing this.

Thanks for the review !  Do I need a second review for those tools like in hotspot ?

**Comment by AlanBateman** [comment]
> > Looks good. Thanks for fixing this.
> 
> Thanks for the review ! Do I need a second review for those tools like in hotspot ?

One reviewer is okay here and Mandy is the expert on this tool. So I think you are good to go.

File: test/langtools/tools/jdeps/optionalDependency/OptionalDependencyTest.java
Comment (AlanBateman) [inline test/langtools/tools/jdeps/optionalDependency/OptionalDependencyTest.java:74]: Just a heads-up that this test is failing in our CI because this needs to use File.pathSeparator rather than ":". Here's the exception:

```
test OptionalDependencyTest.optionalDependenceNotResolved(): failure
java.nio.file.InvalidPathException: Illegal char <:> at index 6: m2.jar:m3.jar
	at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
	at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
	at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:232)
	at java.base/java.nio.file.Path.of(Path.java:147)
	at java.base/java.nio.file.Paths.get(Paths.java:69)
	at jdk.jdeps/com.sun.tools.jdeps.JdepsConfiguration$Builder.createModulePathFinder(JdepsConfiguration.java:582)
	at jdk.jdeps/com.sun.tools.jdeps.JdepsConfiguration$Builder.appModulePath(JdepsConfiguration.java:473)
	at jdk.jdeps/com.sun.tools.jdeps.JdepsTask.buildConfig(JdepsTask.java:584)
	at jdk.jdeps/com.sun.tools.jdeps.JdepsTask.run(JdepsTask.java:558)
	at jdk.jdeps/com.sun.tools.jdeps.JdepsTask.run(JdepsTask.java:534)
	at jdk.jdeps/com.sun.tools.jdeps.Main.run(Main.java:65)
	at jdk.jdeps/com.sun.tools.jdeps.Main$JDepsToolProvider.run(Main.java:78)
	at JdepsRunner.run(JdepsRunner.java:68)
	at OptionalDependencyTest.optionalDependenceNotResolved(OptionalDependencyTest.java:77)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:599)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:174)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:822)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:147)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.testng.TestRunner.privateRun(TestRunner.java:764)
	at org.testng.TestRunner.run(TestRunner.java:585)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:384)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:378)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:337)
	at org.testng.SuiteRunner.run(SuiteRunner.java:286)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1218)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
	at org.testng.TestNG.runSuites(TestNG.java:1069)
	at org.testng.TestNG.run(TestNG.java:1037)
	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:93)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:312)
	at java.base/java.lang.Thread.run(Thread.java:1588)
```

Comment (MBaesken) [reply test/langtools/tools/jdeps/optionalDependency/OptionalDependencyTest.java:74]: Hi Alan, thanks for the heads up. I created https://bugs.openjdk.org/browse/JDK-8294840 .

Comment (mlchung) [reply test/langtools/tools/jdeps/optionalDependency/OptionalDependencyTest.java:74]: @AlanBateman thanks for reporting this.  Sorry I missed this from my review.

**Pull request closed without merging**
